File: app/src/main/java/com/example/johnqualls/recyclerviewdemo/TestAdapter.java
package com.example.johnqualls.recyclerviewdemo;

import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;

import java.util.List;

/**
 * Created by john.qualls on 3/12/2017.
 */

public class TestAdapter extends RecyclerView.Adapter<TestAdapter.ViewHolder> {
    private List<Item> mDataset;

    // Provide a reference to the views for each data item
    // Complex data items may need more than one view per item, and
    // you provide access to all the views for a data item in a view holder
    public static class ViewHolder extends RecyclerView.ViewHolder {
        public LinearLayout mLinearLayout;

        public ViewHolder(LinearLayout v) {
            super(v);
            mLinearLayout = v;
        }
    }

    // Provide a suitable constructor (depends on the kind of dataset)
    public TestAdapter(List<Item> myDataset) {
        mDataset = myDataset;
    }

    // Create new views (invoked by the layout manager)
    @Override
    public TestAdapter.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        // create a new view
        LinearLayout v = (LinearLayout) LayoutInflater.from(parent.getContext())
                .inflate(R.layout.row_view, parent, false);
        // set the view's size, margins, paddings and layout parameters
        //...
        ViewHolder vh = new ViewHolder(v);
        return vh;
    }

    // Replace the contents of a view (invoked by the layout manager)
    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        // - get element from your dataset at this position
        // - replace the contents of the view with that element
        Item item = mDataset.get(position);

        Log.i(TestAdapter.class.getName(), "onBindViewHolder() -> PopulateView");
        ((TextView) holder.mLinearLayout.getChildAt(0)).setText(item.getKey());
        ((TextView) holder.mLinearLayout.getChildAt(1)).setText(item.getValue());
    }

    // Return the size of your dataset (invoked by the layout manager)
    @Override
    public int getItemCount() {
        return mDataset.size();
    }

}

File: app/src/main/java/com/example/johnqualls/recyclerviewdemo/TestAdapter.kt
package com.example.johnqualls.recyclerviewdemo

import androidx.recyclerview.widget.RecyclerView
import android.util.Log
import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.TextView
import androidx.recyclerview.widget.ListAdapter

class TestAdapter(private val viewHolderCallback: (TestAdapter, Int) -> Unit) :
        ListAdapter<Item, TestAdapter.ViewHolder>(DiffUtilCallback()) {

    class ViewHolder(var mLinearLayout: LinearLayout) : RecyclerView.ViewHolder(mLinearLayout)

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): TestAdapter.ViewHolder {
        val v = LayoutInflater.from(parent.context)
                .inflate(R.layout.row_view, parent, false) as LinearLayout
        return ViewHolder(v)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val item = getItem(position)
        val view : TextView = holder.mLinearLayout.getChildAt(0) as TextView
        Log.i(TestAdapter::class.java.name, "onBindViewHolder() -> PopulateView")
        view.text = item.name
        view.setOnClickListener{
            viewHolderCallback(this, position)
        }
    }
}